### ISSUE TITLE
Admin page: delete/purge form data including within a date range

### ISSUE THREAD

**Issue by ebruchez** [orbeon/orbeon-forms#5209]
Following #3597 (and covered in part by #5064), we'd like, on the Admin page, to be able to:

- delete all form data for a given published form (app/form/version)
    - including doing this when deleting the published form definition
    - option to keep (as) historical data?
- delete data within a range, including just "newer than" or "older than"
    - option to keep (as) historical data?

We need:

- at least one or two levels of confirmation!
- to show the user how many "rows" will be deleted (including historical data)

We also would like to be able to delete just historical data:

- all
- date range

The Admin page could also show, at all times, how much data is in the database for each published form definition, in a separate column.

**Comment by ebruchez** [comment]
See also #2886.

**Comment by avernet** [comment]
[+1 from customer](https://3.basecamp.com/3600924/buckets/23878674/messages/5235361464)

**Comment by ebruchez** [comment]
See also Import/export forms with dependencies #779.

**Comment by ebruchez** [comment]
Two implementation strategies:

1. implement purging at the provider level
    - more performant but requires more changes to the provider
2. implement purging at the proxy level
    - less performant but requires less changes to the provider
    - all the filtering logic (dates, etc.) can be done at the proxy level

In any case, the provider will need to use SQL `DELETE` as is the case when deleting drafts.

For attachments:

- if we are purging all data including historical for a document id, then all attachments are deleted
    - fairly simple scenario
- if we are only deleting a subset of the data for a document id, then we must only delete (garbage collect) the attachments that are no longer in use after the purge
    - algorithm
        - read all historical data for the given document id (`/history/` API)
        - extract all attachment paths
        - determine those unused

**Comment by ebruchez** [comment]
Below, "purge" means "force delete".

Here is how things work right now for CUD requests:

- if request is for data AND NOT for attachment (but could be draft or not)
    - SQL `DELETE` all drafts for the given document id in `orbeon_i_control_text`
    - then SQL `DELETE` all drafts for the given document id in
        - `orbeon_form_data_attach`
            - only if `! dataPart.isDraft`, because "when saving the data for a draft, we don't want to remove draft attachments from `orbeon_form_data_attach`, otherwise the attachments which were just saved are immediately removed! So we remove draft attachments from `from orbeon_form_data_attach` only when we are saving data which is not for a draft. Note that there is no garbage collection for attachments."
        - `orbeon_i_current`
        - `orbeon_form_data`
- once relevant drafts are deleted, AND IF we are NOT, specifically, deleting data drafts
    - insert into one
        - `orbeon_form_definition`
        - `orbeon_form_data`
        - `orbeon_form_definition_attach`
        - `orbeon_form_data_attach`
- then data is indexed
    - SQL `DELETE` from `orbeon_i_control_text`/`orbeon_i_current`
    - `INSERT` into `orbeon_i_current`/`orbeon_i_control_text`

For purging:

- Q: Do we still need to delete drafts?
    - A: I think yes, so no change.
- insertion must NOT take place
- indexing must NOT take place

However, we must SQL `DELETE` stuff. We must support:

- HTTP method is `DELETE`, so `delete == true`
- purge of data (not data draft)
- purge of data (not data draft) attachment
- we don't need to handle explicitly deleting drafts, so
    - `dataPart.exists(! _.isDraft)`
- we must support date ranges to some level
    - what makes sense: purging data and historical data *before* a certain date
    - what probably doesn't make sense: purging data and historical data *after* a certain date *unless* it is all contained after that date
        - otherwise you could have possibly some historical data without current data
        - so from this point of view, this would be different from the export

Now what do we need to support at the API level? If we allow force deleting current data, what happens to historical data?

- the expectation can't be that the previous revision becomes suddenly the latest, as the result of a `DELETE` would suddenly be a resource springing to existence
- we could delete all historical data automatically in that case, since we don't need to support only force deleting current data but not historical data

We could, alternatively, NOT support force deleting current data, and require, instead:

- first `DELETE` current data the normal way
    - all data for the document id then becomes historical, including the last one with `deleted = 'Y'`
    - this would also cause the data to become deindexed
- then purge either
    - data for the given document id, which would delete everything in one shot
    - or each revision individually
    - however, we need to support date ranges, so either we should pass that to the provider, or do this at the proxy level and pass the last modified date, which is better for now

What is needed:

- if purge current data (so NOT historical), deindex
    - SQL `DELETE` from `orbeon_i_control_text`/`orbeon_i_current`
    - so implement separate `deindex()` function
    - *unless* we opt for not supporting purging current data, in which case we don't need to deindex as that's done by the regular `DELETE`
- if purge historical data only, do NOT deindex
- if purge data attachment, delete all from attachment table for document id
    - this is because we don't support revision history for attachments
        - we don't have a search API for that so we don't know the modified dates

All the above considered, I am inclined to say:

- don't support purging current data
- support purging historical data
    - always pass a last modified time
- support purging data attachments (current and historical in one shot)
    - never pass a last modified time
- support `data-last-modified-time-lt`
- don't support `data-last-modified-time-gt`

**Comment by ebruchez** [comment]
Purging a given instance of historical data:

- don't touch attachment tables
- delete from `orbeon_form_data` only

Purging a given data attachment:

- delete from `orbeon_form_data_attach` only

**Comment by ebruchez** [comment]
- [x] implement purging in the relational CRUD implementation
- [ ] proxy
    - [ ] purge data
        - call regular delete on current data in date range
        - call purge delete on revision data in date range
    - [ ] attachments GC
- [ ] unit tests
- [ ] doc API
- [ ] doc feature
- [ ] port to master

**Comment by ebruchez** [comment]
Quick and dirty UI mockup:

![image](https://github.com/orbeon/orbeon-forms/assets/105769/ceee4d26-802e-480a-a75a-988e7032e057)

**Comment by ebruchez** [comment]
On attachments GC:

- we don't have the case of "delete all data" in that we have decided to delete each data/revision individually
- so we run `PersistenceApi.dataHistory() in this case
- this calls `processFormData()`
    - we skip reading the data if we are out of the date range
        - this means, as we are iterating the revision history from newer to older, that we might not read data for the first `n` iterations
- however, in the case of purging, we need to have the data so we can collect attachment references
- so we need to add a way to collect attachment references
- right now, we collect attachment references for a given document id so we can avoid reading and storing duplicate references into the zip

**Comment by avernet** [comment]
[+1 from customer](https://3.basecamp.com/3600924/buckets/16915667/messages/6272019565)

**Comment by ebruchez** [comment]
Before we can enable the Purge feature, we need to have comprehensive tests.

We already have infrastructure (see `RestApiTest`) to perform that kind of tests.

The idea is to:

1. Populate the database with published form definitions, form data, and historical form data (revision history). Ideally, there should be a fairly large number of those, including using multiple apps/forms/versions.
2. Perform pruning of this using multiple filters.
3. Check, using APIs (hopefully we don't need to check at the database level) that what needs to be still present in the database is present, and what is expected to have been pruned has been pruned.

We could also at the same time test the Export (#779), and test that the resulting zip file contains what is expected.

**Comment by ebruchez** [comment]
Admin user might not have permissions, but also might not even receive search results for documents where she has no permissions at all.

We should detect that case and warn the user in case documents cannot be deleted.

**Comment by obruchez** [comment]
Tests to add to PurgeTest:

- [ ] purge with date ranges
- [ ] purge with multiple matches
- [ ] purge with attachments
- [ ] purge with multiple applications (note that the application is currently used to test different persistence providers)
- [ ] other important cases (check with @ebruchez)
- [ ] check that some parameters are correctly rejected (e.g. match=*/*/2, etc.)

The export feature is currently used / indirectly tested in PurgeTest, but more tests should be implemented:

- [ ] add ExportTest
- [ ] test important cases
- [ ] check actual content of exported form definitions/data (MD5?)

getProviders has been modified (case with app specified but form not specified):

- [ ] add test for getProviders

Nice to have:

- [ ] publish the form definitions via the API instead of hard-coding the metadata/migration element content

**Comment by ebruchez** [comment]
Created new issues for follow-up tasks and closing this.
